mixnet/mixnet.py

```diff
@@ -4,44 +4,59 @@
 from dataclasses import dataclass
 from typing import List
 
-from mixnet.fisheryates import FisherYates
 from mixnet.node import MixNode
 
 
-@dataclass
 class Mixnet:
-    mix_nodes: List[MixNode]
-
-    # Build a new topology deterministically using an entropy.
-    # The entropy is expected to be injected from outside.
-    #
-    # TODO: Implement constructing a new topology in advance to minimize the topology transition time.
-    #       https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#9a7f6089e210454bb11fe1c10fceff68
-    def build_topology(
-        self,
-        entropy: bytes,
-        n_layers: int,
-        n_nodes_per_layer: int,
-    ) -> MixnetTopology:
-        num_nodes = n_nodes_per_layer * n_layers
-        assert num_nodes < len(self.mix_nodes)
-
-        shuffled = FisherYates.shuffle(self.mix_nodes, entropy)
-        sampled = shuffled[:num_nodes]
-        layers = []
-        for l in range(n_layers):
-            start = l * n_nodes_per_layer
-            layer = sampled[start : start + n_nodes_per_layer]
-            layers.append(layer)
-        return MixnetTopology(layers)
-
-    def choose_mixnode(self) -> MixNode:
-        return random.choice(self.mix_nodes)
+    __topology: MixnetTopology | None = None
+
+    @property
+    def topology(self) -> MixnetTopology:
+        if self.__topology is None:
+            raise RuntimeError("topology is not set yet")
+        return self.__topology
+
+    @topology.setter
+    def topology(self, topology: MixnetTopology) -> None:
+        """
+        Replace the old topology with the new topology received, and start establishing new network connections in background.
+
+        In real implementations, this method should be a long-running task, accepting topologies periodically.
+        Here in the spec, this method has been simplified as a setter, assuming the single-thread test environment.
+        """
+        self.__topology = topology
+        self.__establish_connections()
+
+    def __establish_connections(self) -> None:
+        """
+        Establish network connections in advance based on the topology received.
+
+        This is just a preparation to forward subsequent packets as quickly as possible,
+        but this is not a strict requirement.
+
+        In real implementations, this should be a background task.
+        """
+        pass
 
 
 @dataclass
 class MixnetTopology:
+    # In production, this can be a 1-D array, which is accessible by indexes.
+    # Here, we use a 2-D array for readability.
     layers: List[List[MixNode]]
 
     def generate_route(self) -> list[MixNode]:
         return [random.choice(layer) for layer in self.layers]
+
+    def choose_mix_destionation(self) -> MixNode:
+        all_mixnodes = [mixnode for layer in self.layers for mixnode in layer]
+        return random.choice(all_mixnodes)
```

Comment:
In this code, yes, the destination is selected by the mixnet layer automatically within the set of all mix nodes in the topology. It’s because the only information that the mixnet layer is the topology.
But I just remembered that we’ve discussed to choose a destination from the full set of nodes (including nodes not selected as mix). 
The robustness layer knows the full set. But, when the consensus layer or other layers try to send a message using mixnet, I think they will directly interact with the mixnet layer without asking the robustness layer.

Actually, after writing this Python code, now I think we can actually use mixnodes in the exit layer as mix destination, as you mentioned at the end of your comment. That will save one hop. Design is gonna be simpler. And, I don’t see any downside yet in terms of security

Reply:
If we use the exit layer as mix destinations, maybe adversaries may know that all packets coming into a node F in the exit layer is associated with some of them. That could be a downside if there is not much traffic in the network. But I think this information can be also leaked if someone can track egresses of the exit layer. Also in Nym, they have the concept of ‘service provider’ where all associated packets are gathered to. Therefore, I agree with this strategy. If you think so, I’ll update this and Notion

Reply:
> Therefore, I agree with this strategy. If you think so, I’ll update this and Notion

@madxor I've thought about it. I think it's reasonable to use the last mix layer for mix destinations, instead of introducing an extra hop. I just updated [Notion](https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#ebb923fa6a7443a0a36737906104343c), and Python code: 2aa9d5d. I added a new [diagram](https://www.notion.so/Mixnet-Specification-807b624444a54a4b88afa1cc80e100c2?pvs=4#ebb923fa6a7443a0a36737906104343c) in Notion, so anyone can understand it clearly.